### PR TITLE
[fix] Handle raw image bytes in workflow step _convert_image_artifacts_to_images

### LINKED ISSUE #6709: fix: Workflow step silently drops images - _convert_image_artifacts_to_images assumes bytes are base64
## Bug Description

When an image is sent to a Workflow (via Telegram, WhatsApp, or any interface), the `_convert_image_artifacts_to_images` method in `libs/agno/agno/workflow/step.py` (line 1589) tries to decode raw image bytes as UTF-8 text. This fails because raw JPEG/PNG bytes aren't valid UTF-8. The error is caught silently and the image is just skipped — the agent never receives it.

## Reproduction

1. Connect any interface (Telegram, WhatsApp, Slack) to a Workflow
2. Send an image to the bot
3. The image is silently dropped — the agent responds as if no image was sent
4. Server logs show: `Failed to process image content: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte`

## Root Cause

Line 1589 in `workflow/step.py`:
```python
if isinstance(img_artifact.content, bytes):
    base64_str: str = img_artifact.content.decode("utf-8")  # Crashes on raw image bytes
    actual_image_bytes = base64.b64decode(base64_str)
```

Every image producer in the codebase stores raw bytes, not base64-encoded bytes — so this always fails.

## Suggested Fix

```python
if isinstance(img_artifact.content, bytes):
    try:
        base64_str = img_artifact.content.decode("utf-8")
        actual_image_bytes = base64.b64decode(base64_str)
    except (UnicodeDecodeError, Exception):
        actual_image_bytes = img_artifact.content  # Raw image bytes, use directly
```

## Impact

All images passed to Workflows from any source are silently dropped during step execution.

Found while testing PR #6377 (Telegram interface).

### PR DESCRIPTION
**Pull request by Himanshu040604**
## Summary

When an image is sent to a Workflow (via Telegram, WhatsApp, Slack, or any interface), the `_convert_image_artifacts_to_images` method in `libs/agno/agno/workflow/step.py` tries to decode raw image bytes as UTF-8 text. This fails because raw JPEG/PNG bytes aren't valid UTF-8. The error is caught silently and the image is skipped — the agent never receives it.

The fix wraps the UTF-8 decode in a try/except and falls back to using the raw bytes directly.

fixes #6709

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [x] Self-review completed
- [x] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [x] Tested in clean environment
- [x] Tests added/updated (if applicable)

---

## Additional Notes

- Every image producer in the codebase stores raw bytes, not base64 (OpenAI tools, Gemini tools, Telegram, WhatsApp, Slack routers)
- No upstream transformation converts raw bytes to base64 before reaching this method
- All 259 existing workflow unit tests pass with this change
- Verified with 4 manual tests: raw JPEG, raw PNG, base64-encoded bytes (no regression), and URL-based images

### CODE REVIEW

File: libs/agno/agno/workflow/step.py [hunk 1]
```diff
@@ -1817,9 +1817,13 @@ def _convert_image_artifacts_to_images(self, image_artifacts: List[Image]) -> Li
...
                     # Try to decode as base64 first (for images from OpenAI tools)
                     if isinstance(img_artifact.content, bytes):
-                        # Decode bytes to string, then decode base64 to get actual image bytes
-                        base64_str: str = img_artifact.content.decode("utf-8")
-                        actual_image_bytes = base64.b64decode(base64_str)
+                        try:
+                            # Attempt UTF-8 decode in case bytes are base64-encoded text
+                            base64_str: str = img_artifact.content.decode("utf-8")
+                            actual_image_bytes = base64.b64decode(base64_str)
+                        except (UnicodeDecodeError, Exception):
```
Comment (kausmeows) [inline libs/agno/agno/workflow/step.py]: can we pls use `binascii.Error` instead of generic Exception for the base64 decode failure, as that's what `base64.b64decode` raises for invalid input.

Comment (Himanshu040604) [reply libs/agno/agno/workflow/step.py]: Done! Updated to use `binascii.Error` instead of generic `Exception`, and added `import binascii` alongside the existing `import base64`. See commit b883a01.

[the remark this answers is not in the document]
Comment (Himanshu040604) [reply libs/agno/agno/workflow/step.py]: Good catch — addressed in b883a01. Now uses `except (UnicodeDecodeError, binascii.Error)` which covers both specific failure modes without being overly broad.

File: libs/agno/agno/workflow/step.py [hunk 2]
```diff
@@ -1817,9 +1817,13 @@ def _convert_image_artifacts_to_images(self, image_artifacts: List[Image]) -> Li
                 try:
                     # Try to decode as base64 first (for images from OpenAI tools)
                     if isinstance(img_artifact.content, bytes):
-                        # Decode bytes to string, then decode base64 to get actual image bytes
-                        base64_str: str = img_artifact.content.decode("utf-8")
-                        actual_image_bytes = base64.b64decode(base64_str)
+                        try:
+                            # Attempt UTF-8 decode in case bytes are base64-encoded text
+                            base64_str: str = img_artifact.content.decode("utf-8")
+                            actual_image_bytes = base64.b64decode(base64_str)
```
[the remark this answers is not in the document]
Comment (Himanshu040604) [reply libs/agno/agno/workflow/step.py:1824]: Added 7 unit tests in `libs/agno/tests/unit/workflow/test_step_image_conversion.py` covering:
- Raw JPEG bytes preserved (not dropped)
- Raw PNG bytes preserved (not dropped)
- Base64-encoded bytes still decoded correctly (no regression)
- URL-based images pass through unchanged
- Empty input returns empty output
- Multiple mixed images all preserved
- Image with mime_type handled correctly

All 266 workflow tests pass (259 existing + 7 new). See commit b883a01.

**Comment by Himanshu040604** [comment]
Hi @kausmeows, thanks for the review! All feedback has been addressed in commit b883a01:

**1. `binascii.Error` instead of generic `Exception`** — Done. Now uses `except (UnicodeDecodeError, binascii.Error)` with `import binascii` added.

**2. Cross-check for audio/video** — Investigated both:
- `_convert_audio_artifacts_to_audio` (line ~1784): passes `content` directly to `Audio(content=...)` without any `.decode("utf-8")` call — **not affected**.
- `_convert_video_artifacts_to_videos` (line ~1853): passes `content` directly to `Video(content=...)` without any `.decode("utf-8")` call — **not affected**.

The bug was specific to `_convert_image_artifacts_to_images` because it's the only method that attempts UTF-8 decode + base64 decode on the content bytes.

**3. Tests added** — 7 unit tests in `libs/agno/tests/unit/workflow/test_step_image_conversion.py`:
- `test_raw_jpeg_bytes_not_dropped` — raw JPEG bytes preserved
- `test_raw_png_bytes_not_dropped` — raw PNG bytes preserved
- `test_base64_encoded_bytes_decoded` — base64 path still works (no regression)
- `test_url_image_passes_through` — URL images unchanged
- `test_empty_list_returns_empty` — edge case
- `test_multiple_images_all_preserved` — mixed types
- `test_image_with_mime_type_sets_format` — mime_type handling

All 266 workflow tests pass (259 existing + 7 new).

**Comment by Himanshu040604** [comment]
Hey @kausmeows, addressed all 3 points in commit b883a01:

1. **binascii.Error** — switched from generic `Exception`, added `import binascii`

2. **Audio/video check** — both `_convert_audio_artifacts_to_audio` and `_convert_video_artifacts_to_videos` pass content directly without `.decode("utf-8")`, so they're not affected. This bug was image-specific.

3. **Tests** — added 7 unit tests covering raw JPEG/PNG bytes, base64 decoding, URL passthrough, empty input, multiple images, and mime_type. All 266 workflow tests pass.
